src/ol/render/webgl/BatchRenderer.js

```diff
@@ -0,0 +1,201 @@
+/**
+ * @module ol/render/webgl/BatchRenderer
+ */
+import {WebGLWorkerMessageType} from './constants.js';
+import {abstract} from '../../util.js';
+import {
+  create as createTransform,
+  makeInverse as makeInverseTransform,
+  multiply as multiplyTransform,
+  translate as translateTransform,
+} from '../../transform.js';
+
+/**
+ * @typedef {Object} CustomAttribute A description of a custom attribute to be passed on to the GPU, with a value different
+ * for each feature.
+ * @property {string} name Attribute name.
+ * @property {function(import("../../Feature").default, Object<string, *>):number} callback This callback computes the numerical value of the
+ * attribute for a given feature (properties are available as 2nd arg for quicker access).
+ */
+
+let workerMessageCounter = 0;
+
+/**
+ * @classdesc Abstract class for batch renderers.
+ * Batch renderers are meant to render the geometries contained in a {@link module:ol/render/webgl/GeometryBatch}
+ * instance. They are responsible for generating render instructions and transforming them into WebGL buffers.
+ */
+class AbstractBatchRenderer {
+  /**
+   * @param {import("../../webgl/Helper.js").default} helper WebGL helper instance
+   * @param {Worker} worker WebGL worker instance
+   * @param {string} vertexShader Vertex shader
+   * @param {string} fragmentShader Fragment shader
+   * @param {Array<CustomAttribute>} customAttributes List of custom attributes
+   */
+  constructor(helper, worker, vertexShader, fragmentShader, customAttributes) {
+    /**
+     * @type {import("../../webgl/Helper.js").default}
+     * @protected
+     */
+    this.helper_ = helper;
+
+    /**
+     * @type {Worker}
+     * @protected
+     */
+    this.worker_ = worker;
+
+    /**
+     * @type {WebGLProgram}
+     * @protected
+     */
+    this.program_ = this.helper_.getProgram(fragmentShader, vertexShader);
+
+    /**
+     * A list of attributes used by the renderer.
+     * @type {Array<import('../../webgl/Helper.js').AttributeDescription>}
+     * @protected
+     */
+    this.attributes_ = [];
+
+    /**
+     * @type {Array<CustomAttribute>}
+     * @protected
+     */
+    this.customAttributes_ = customAttributes;
```

Comment:
Since this is used elsewhere, it should not be named with a `_` suffix.  I pushed an additional commit to change this.